File: front/src/services/category.ts
// services/category.ts
import api from "./api";
import { CategoryCreate, CategoryResponse, CategoryUpdate } from "@/types/category";

// Cria uma categoria (exige permissões de admin no backend)
export const createCategory = async (category: CategoryCreate): Promise<CategoryResponse> => {
  const response = await api.post("/categories/", category);
  return response.data;
};

// Lista todas as categorias
export const listCategories = async (): Promise<CategoryResponse[]> => {
  const response = await api.get("/categories/");
  return response.data;
};

// Obtém uma categoria pelo ID
export const getCategory = async (categoryId: number): Promise<CategoryResponse> => {
  const response = await api.get(`/categories/${categoryId}`);
  return response.data;
};

// Atualiza os dados de uma categoria (exige permissões de admin)
export const updateCategory = async (
  categoryId: number,
  categoryData: CategoryUpdate
): Promise<CategoryResponse> => {
  const response = await api.put(`/categories/${categoryId}`, categoryData);
  return response.data;
};

// Exclui uma categoria (exige permissões de admin)
export const deleteCategory = async (categoryId: number): Promise<void> => {
  await api.delete(`/categories/${categoryId}`);
};

File: back/app/routers/category.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List

from app.database import get_db
from app.schemas.category import CategoryCreate, CategoryUpdate, CategoryResponse
from app.models.category import Category
from app.dependencies import get_current_admin, get_current_user

router = APIRouter(
    tags=["Categories"]
)

@router.post("/", response_model=CategoryResponse, dependencies=[Depends(get_current_admin)])
def create_category(category: CategoryCreate, db: Session = Depends(get_db)):
    # Verifica se a categoria já existe
    existing = db.query(Category).filter(Category.name == category.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="Categoria já existe")
    new_category = Category(name=category.name)
    db.add(new_category)
    db.commit()
    db.refresh(new_category)
    return new_category

@router.get("/", response_model=List[CategoryResponse])
def list_categories(db: Session = Depends(get_db), current_user = Depends(get_current_user)):
    categories = db.query(Category).all()
    return categories

@router.get("/{category_id}", response_model=CategoryResponse)
def get_category(category_id: int, db: Session = Depends(get_db), current_user = Depends(get_current_user)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Categoria não encontrada")
    return category

@router.put("/{category_id}", response_model=CategoryResponse, dependencies=[Depends(get_current_admin)])
def update_category(category_id: int, category_data: CategoryUpdate, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Categoria não encontrada")
    if category_data.name is not None:
        category.name = category_data.name
    db.commit()
    db.refresh(category)
    return category

@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT, dependencies=[Depends(get_current_admin)])
def delete_category(category_id: int, db: Session = Depends(get_db)):
    category = db.query(Category).filter(Category.id == category_id).first()
    if not category:
        raise HTTPException(status_code=404, detail="Categoria não encontrada")
    db.delete(category)
    db.commit()
    return
